### PR TITLE
Sec providers android

### PR DESCRIPTION
**Pull request by commjoen**
This is a first fix for https://github.com/OWASP/owasp-mstg/issues/517

### CODE REVIEW

File: Document/0x05g-Testing-Network-Communication.md [hunk 1]
```diff
@@ -196,3 +196,180 @@ The SSL pinning process should be implemented as described on the static analysi
...
+- Run the application in debugmode, then make a debug-point right at the point where the app will make it first contact with the backend.
+- rightclick at the code that is highlighted and select `Evaluate Expression`
+- Type `Security.getProviders()` and press enter
+- Check the providers and see if you can find `GmsCore_OpenSSL` which should be the new toplisted provider.
+
+When you do not have the source-code:
+- Use Xposed to hook into `java.security` package, then hook into `java.security.Security` with the method `getProviders` with no arguments. The return value is an Array of `Provider`. 
+- Check if the first provider is `GmsCore_OpenSSL`.
+
+--- {TODO: ADD ANDROID NATIVE DYNAMIC ANALYSIS}
```
Comment (bolot) [inline Document/0x05g-Testing-Network-Communication.md]: Per IRL discussion, perhaps leave the discussion at a high level. Add a reference to the chapter that discusses general secure networking principles. There may be high variability in native/NDK libraries.

File: Document/0x05g-Testing-Network-Communication.md [hunk 2]
```diff
@@ -196,3 +196,180 @@ The SSL pinning process should be implemented as described on the static analysi
...
+
+
+### Testing used connectivity and communication libraries
+
+#### Overview 
+Android relies on a security provider to provide SSL/TLS based connections. The problem with this security provider (for instance OpenSSL) which is packed with the device, is that it often has bugs and/or vulnerabilities<sup>[1]</sup>.
+Developers need to make sure that the application will install a proper security provider to make sure that there will be lesser bugs and vulnerabilities.
+
+#### Static Analysis
+In case of an Android SDK based application. The application should have a dependency on the GooglePlayServices. (e.g. in a. gradle build file, you will find `compile 'com.google.android.gms:play-services-gcm:x.x.x'` in the dependencies block). Next you need to make sure that the `ProviderInstaller` class is called with either `installIfNeeded()` or with `installIfNeededAsync()` is called as soon as possible and that errors comming from that are handled correctly: either the application informs the API on his lesser secure state or it restricts the user in its possible actions as all https-traffic should now be deemed more risky.
```
Comment (bolot) [inline Document/0x05g-Testing-Network-Communication.md]: May want to add a small clarification about "as soon as possible", maybe a code snippet, e.g., "this should be a static code block".

File: Document/0x05g-Testing-Network-Communication.md [hunk 3]
```diff
@@ -196,3 +196,180 @@ The SSL pinning process should be implemented as described on the static analysi
...
+
+--- {TODO: ADD ANDROID NATIVE STATIC ANALYSIS}
+
+#### Dynamic Analysis
+When you have the source-code: 
+
+- Run the application in debugmode, then make a debug-point right at the point where the app will make it first contact with the backend.
```
Comment (bolot) [inline Document/0x05g-Testing-Network-Communication.md]: Words got merged: "debug mode".
Debug-point - probably should be "breakpoint".
"...right where the app makes its first connection to the backend"?

File: Document/0x05g-Testing-Network-Communication.md [hunk 4]
```diff
@@ -196,3 +196,180 @@ The SSL pinning process should be implemented as described on the static analysi
...
+- rightclick at the code that is highlighted and select `Evaluate Expression`
```
Comment (bolot) [inline Document/0x05g-Testing-Network-Communication.md]: "Right click" missing a space.

File: Document/0x05g-Testing-Network-Communication.md (hunk 2 above)
Comment (bolot) [inline Document/0x05g-Testing-Network-Communication.md]: Oh, I see you have the example later in "Remediation".

File: Document/0x05g-Testing-Network-Communication.md [hunk 5]
```diff
@@ -196,3 +196,180 @@ The SSL pinning process should be implemented as described on the static analysi
...
+- Use Xposed to hook into `java.security` package, then hook into `java.security.Security` with the method `getProviders` with no arguments. The return value is an Array of `Provider`. 
+- Check if the first provider is `GmsCore_OpenSSL`.
+
+--- {TODO: ADD ANDROID NATIVE DYNAMIC ANALYSIS}
+
+#### Remediation
+To make sure that the application is using a patched security provider, the application needs to use the `ProviderInstaller` class which comes with the Google Play services. The Google Play Services can be installed as a dependency in the build.gradle file by adding `compile 'com.google.android.gms:play-services-gcm:x.y.z'` (where x.y.z is a version number) in the dependencies block.
+Next, the `ProviderInstaller` needs to be called as early as possible by a component of the application. Here are two adjusted examples from google on how this could work. In both cases, the developer needs to handle the exceptions propperly and it might be wise to report to the backend when the application is working with an unpatched security provider. The first example shows how to do the installation synchronously, the second example shows how to do it asynchronously.
```
Comment (bolot) [inline Document/0x05g-Testing-Network-Communication.md]: "Google" should be capitalized.
Properly typo – double p.

File: Document/0x05e-Testing-Cryptography.md [hunk 6]
```diff
@@ -6,11 +6,26 @@
...
+    builder.append(provider.getName())
+            .append(" ")
+            .append(provider.getVersion())
+            .append("(")
+            .append(provider.getInfo())
+            .append(")\n");
+}
+String providers = builder.toString();
+//now display the string on the screen or in the logs for debugging.
+```
```
Comment (bolot) [inline Document/0x05e-Testing-Cryptography.md]: Might be useful to add a short, 1-3 line example of the output, along with the device/OS info, e.g., "Below is an excerpt of the output from Nexus 6P running Android 7.1.1, security patch level June 2017"

File: Document/0x05g-Testing-Network-Communication.md (hunk 2 above)
Comment (commjoen) [inline Document/0x05g-Testing-Network-Communication.md]: using a static block might not help as you rather want to adhere to the lifecycle of things. But i tried to be more elaborate and referred to the remediation.

File: Document/0x05g-Testing-Network-Communication.md (hunk 3 above)
Comment (commjoen) [inline Document/0x05g-Testing-Network-Communication.md]: you're totally right! Thanks!

File: Document/0x05g-Testing-Network-Communication.md (hunk 4 above)
Comment (commjoen) [inline Document/0x05g-Testing-Network-Communication.md]: 👍

File: Document/0x05g-Testing-Network-Communication.md (hunk 3 above)
Comment (sushi2k) [inline Document/0x05g-Testing-Network-Communication.md]: Agree, should be breakpoint. I also do not understand: what you mean with backend in this context?

File: Document/0x05g-Testing-Network-Communication.md (hunk 1 above)
Comment (commjoen) [inline Document/0x05g-Testing-Network-Communication.md]: Thanks! Will leave specific analysis out for now.

File: Document/0x05g-Testing-Network-Communication.md (hunk 5 above)
Comment (commjoen) [inline Document/0x05g-Testing-Network-Communication.md]: 👍  fixed it :)

File: Document/0x05e-Testing-Cryptography.md (hunk 6 above)
Comment (commjoen) [inline Document/0x05e-Testing-Cryptography.md]: Fixed it :)

File: Document/0x05g-Testing-Network-Communication.md (hunk 3 above)
Comment (commjoen) [inline Document/0x05g-Testing-Network-Communication.md]: fixed that too :)

**Pull request merged**